pghoard/rohmu/delta/snapshot.py

```diff
@@ -0,0 +1,231 @@
+
+        The output to outside is just root object's hash, as well as list
+        of other hashes which correspond to files referred to within the
+        file list contained in root object.
+
+        Note that any call to public API MUST be made with
+        snapshotter.lock held. This is because Snapshotter is process-wide
+        utility that is shared across operations, possibly used from
+        multiple threads, and the single-operation-only mode of operation
+        is not exactly flawless (the 'new operation can be started with
+        old running' is intentional feature but new operation should
+        eventually replace the old). The lock itself might not need to be
+        built-in to Snapshotter, but having it there enables asserting its
+        state during public API calls.
+        """
+    def __init__(self, *, src, dst, globs, src_iterate_func: Optional[Callable] = None, parallel=1):
+        assert globs
+        self.src = Path(src)
+        self.dst = Path(dst)
+        self.globs = globs
+        self.src_iterate_func = src_iterate_func
+        self.relative_path_to_snapshotfile = {}
+        self.hexdigest_to_snapshotfiles = {}
+        self.parallel = parallel
+        self.lock = threading.Lock()
+        self.empty_dirs = []
+
+    def _list_files(self, basepath: Path):
+        result_files = set()
+        for glob in self.globs:
+            for path in basepath.glob(glob):
+                if not path.is_file() or path.is_symlink():
+                    continue
+                relpath = path.relative_to(basepath)
+                result_files.add(relpath)
+
+        return sorted(result_files)
+
+    def _list_dirs_and_files(self, basepath: Path):
+        files = self._list_files(basepath)
+        dirs = {p.parent for p in files}
+        return sorted(dirs), files
+
+    def _add_snapshotfile(self, snapshotfile: SnapshotFile):
+        old_snapshotfile = self.relative_path_to_snapshotfile.get(snapshotfile.relative_path, None)
+        if old_snapshotfile:
+            self._remove_snapshotfile(old_snapshotfile)
+        self.relative_path_to_snapshotfile[snapshotfile.relative_path] = snapshotfile
+        if snapshotfile.hexdigest:
+            self.hexdigest_to_snapshotfiles.setdefault(snapshotfile.hexdigest, []).append(snapshotfile)
+
+    def _remove_snapshotfile(self, snapshotfile: SnapshotFile):
+        assert self.relative_path_to_snapshotfile[snapshotfile.relative_path] == snapshotfile
+        del self.relative_path_to_snapshotfile[snapshotfile.relative_path]
+        if snapshotfile.hexdigest:
+            self.hexdigest_to_snapshotfiles[snapshotfile.hexdigest].remove(snapshotfile)
+
+    def _snapshotfile_from_path(self, relative_path):
+        src_path = self.src / relative_path
+        st = src_path.stat()
+        return SnapshotFile(relative_path=relative_path, mtime_ns=st.st_mtime_ns, file_size=st.st_size)
+
+    def _get_snapshot_hash_list(self, relative_paths):
```

Comment:
nit: maybe not `_get...list` but `_gen_snapshot_hashes` to make clear that it returns not `list` but `generator`?